vortex-array/src/compute/filter.rs

```diff
@@ -103,209 +105,426 @@ fn filter_impl(array: &ArrayData, mask: FilterMask) -> VortexResult<ArrayData> {
+    pub fn slice(&self, offset: usize, length: usize) -> Self {
+        if self.true_count() == 0 {
+            return Self::new_false(length);
+        }
+        if self.true_count() == self.len() {
+            return Self::new_true(length);
+        }
+
+        if let Some(buffer) = self.0.buffer.get() {
+            return Self::from_buffer(buffer.slice(offset, length));
+        }
+
+        let end = offset + length;
+
+        if let Some(indices) = self.0.indices.get() {
+            let indices = indices
+                .iter()
+                .copied()
+                .filter(|&idx| offset <= idx && idx < end)
+                .map(|idx| idx - offset)
+                .collect();
+            return Self::from_indices(length, indices);
+        }
+
+        if let Some(slices) = self.0.slices.get() {
+            let slices = slices
+                .iter()
+                .copied()
+                .filter(|(s, e)| *s < end && *e > offset)
+                .map(|(s, e)| (s.max(offset), e.min(end)))
+                .collect();
+            return Self::from_slices(length, slices);
+        }
+
+        vortex_panic!("No mask representation found")
     }
+}
+
+pub enum FilterIter<'a> {
+    /// Slice of pre-cached indices of a filter mask.
+    Indices(&'a [usize]),
+    /// Slice of pre-cached slices of a filter mask.
+    Slices(&'a [(usize, usize)]),
+}
 
-    #[deprecated(note = "Move to using iter() instead")]
-    pub fn iter_indices(&self) -> VortexResult<impl Iterator<Item = usize> + '_> {
-        Ok(self.boolean_buffer()?.set_indices())
+impl PartialEq for FilterMask {
+    fn eq(&self, other: &Self) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        if self.true_count() != other.true_count() {
+            return false;
+        }
+
+        // Since the true counts are the same, a full or empty mask is equal to the other mask.
+        if self.true_count() == 0 || self.true_count() == self.len() {
+            return true;
+        }
+
+        // Compare the buffer if both masks are non-empty.
+        if let (Some(buffer), Some(other)) = (self.0.buffer.get(), other.0.buffer.get()) {
+            return buffer == other;
+        }
+
+        // Compare the indices if both masks are non-empty.
+        if let (Some(indices), Some(other)) = (self.0.indices.get(), other.0.indices.get()) {
+            return indices == other;
+        }
+
+        // Compare the slices if both masks are non-empty.
+        if let (Some(slices), Some(other)) = (self.0.slices.get(), other.0.slices.get()) {
+            return slices == other;
+        }
+
+        // Otherwise, we fall back to comparison based on sparsity.
+        // We could go further an exhaustively check whose OnceLocks are initialized, but that's
+        // probably not worth the effort.
+        self.boolean_buffer() == other.boolean_buffer()
+    }
+}
+
+impl Eq for FilterMask {}
+
+impl BitAnd for FilterMask {
```

Comment:
how about we define this for `&FilterMask` and `FilterMask` impl would just ref self?